Navigation: Remove blobs that look like a loading state

jasmussen opened the pull request:
## Description

Alternative to #36858 and #37144. Redesigns the navigation block setup state to not show fake menu item skeleton indicators.

The primary problem to solve is to make the navigation block look _unconfigured_, but still in this state blend a little bit into the theme to provide a better initial experience. The problem with the gray skeleton indicators was that they looked like a loading state, making people think the no further action was taken. A secondary issue was that the previous setup state caused layout shifts on selection. Some additional context from the original issue:

> The initial idea was to let themes preinsert a navigation block at an appropriate place in the theme, but leave it otherwise unconfigured. With that goal in mind, the gray blobs were meant to help the placeholder state pass the "squint test", to look at least a little bit like the end result at least when the block was unselected.
> 
> Feedback suggested this didn't quite work as intended, though: the blobs looked like a "skeleton loading state", that they would be replaced with the real menu once loading had completed.
> 
> There's persistence in the navigation block now, with contents saved separately, and options for providing defaults and fallbacks. Depending on where that lands, we might be able to provide better first run experiences. In any case, it changes the equation from before, where an empty and unconfigured navigation block was more likely to appear in themes.

This PR takes a different direction, leaning into the "generic placeholder" look we recently employed for Site Logo and Featured Image:

![navigation](https://user-images.githubusercontent.com/1204802/151158231-80714024-d3ed-4b6e-b1b4-19b1b7aa16ff.gif)

Let me know your thoughts!

## How has this been tested?

Insert a fresh navigation block, select and deselect it.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Navigation: Remove blobs that look like a loading state

priethor commented on packages/block-library/src/navigation/edit/placeholder/placeholder-preview.js:
```diff
@@ -6,24 +6,18 @@ import classnames from 'classnames';
-				<Icon icon={ search } />
-			</li>
-		</ul>
+			<Icon icon={ navigation } />
```
The compass icon can't be seen with darker backgrounds, which is something to consider given TT2:
![image](https://user-images.githubusercontent.com/27339341/144612420-a537e3c5-3eea-4bc6-b13b-7c59ed13e542.png)

priethor reviewed (commented):
Thanks for this new approach, I think it's a good compromise between a lean placeholder and the setup menu when selected.

jasmussen commented:
Oh that's a bug, let me fix that real quick.

jasmussen commented:
Thanks Hector!

@annezazu I know you reported many of the problems with the blobs in the past. Do you have any thoughts to share? Thank you for your time!

annezazu commented:
Thanks for tagging me into this! Compared to the other exploration, I like the consistency and simplicity this provides. I think it will be important for these placeholder states to be thought of cohesively so am in favor of this exploration over on that basis alone: https://github.com/WordPress/gutenberg/pull/36858 

_For great context and thoughts..._ Just this week, I got feedback from someone in the current call for testing around how they expected to see the three blobs on the front end of their site but I haven't heard this as widespread feedback. If anything, my main concern would be around ensuring that the placeholder makes it clear that it's both related to a menu and that action needs to be taken to configure it, especially since compared to the post featured image one does need to take action (vs the featured image just being pulled in).

In looking at both the post featured image and navigation block placeholders together, my brain immediately thinks about how well these placeholders will scale if they all look so very empty. In their current state, it's hard to see at a glance what each of these items are without requiring folks to click into them. I wonder what else could be included instead of just the icon for the navigation block in the placeholder? I don't know if that's going to be enough to differentiate compared to what is in place for featured image:

<img width="672" alt="Screen Shot 2021-12-03 at 12 07 06 PM" src="https://user-images.githubusercontent.com/26996883/144643479-aa9ab22f-cb5a-4632-8c92-3df2384a5c66.png">

vs

<img width="721" alt="Screen Shot 2021-12-03 at 12 06 38 PM" src="https://user-images.githubusercontent.com/26996883/144643413-81473cf7-6469-4e8b-8275-05a7f470de9b.png">

The former really gives more insight at a glance as to what each placeholder represents IMO and I wonder what will be lost by removing some indication that it's for the nav block. Perhaps keep the name in there? 

<img width="755" alt="Screen Shot 2021-12-03 at 12 15 40 PM" src="https://user-images.githubusercontent.com/26996883/144644733-cbdd732c-b2a7-441a-b716-d25065e2c457.png">

jasmussen commented:
Excellent thoughts, thank you all. 

If there's one take-away from parsing this feedback over the weekend, it only affirms that the blobs don't quite work. I think this dashed version is probably _better_, but maybe it isn't that much better. It might need more design thought. And so I think it boils down to the question: do we want to land this for 5.9 and then continue to iterate, or do we want to go with the blobs? 

It's very easy to get caught up in the moment of wanting to pack as much as possible into the release. At the same time, there are more important fixes to make, and there will be the plugin to ship fixes in, and 6.0 a few months later. I'll keep thinking and see if there's a way to thread the needle between _looking like navigation_ and _looking like it needs further action_.

jasmussen committed:
Polish block indicator to include label

jasmussen commented:
I've rebased, refreshed, and updated this one based on feedback. I've also included a new description with a fresh GIF of how it looks.

I think with the dashed lines, and the icon plus label, this placeholder state well balances the need to clearly look like a placeholder, but at the same time look unconfigured.

priethor reviewed (commented):
I love this new iteration, I think it represents a good balance as you point out 👍 

After testing it with a dark background, the dark icon bug has come back:
![image](https://user-images.githubusercontent.com/27339341/151234145-e022c133-d041-4124-ae0b-af68ac4063dd.png)

jasmussen committed:
Inherit color on SVG.

jasmussen commented:
Ack, and I had it fixed before even! Here it is again, thank you:
<img width="676" alt="Screenshot 2022-01-26 at 21 18 38" src="https://user-images.githubusercontent.com/1204802/151240419-91871d73-55f1-4f95-8387-57540a97b777.png">

getdave reviewed (commented):
Thank you for pursing this 👏 

I greatly prefer this state to the blobs. It's much more visual "stable" and it's relatively clear that the Navigation is in an unconfigured state.

I found another bug with the colors on a dark background. I believe you've fixed the placeholder preview but not the placeholder component itself.

https://user-images.githubusercontent.com/444434/151339930-ce3db00d-a128-4ec1-bd55-2260657290dc.mov

Code to fix is here:

https://github.com/WordPress/gutenberg/blob/b3e57fb20d9ebdb87e7644a58314a0191e66aeb5/packages/block-library/src/navigation/edit/placeholder/index.js#L160-L163

This points to redundancy in the code which could be refactored to avoid the need to update in x2 places. Maybe one to add to the list for https://github.com/WordPress/gutenberg/issues/37190

jasmussen commented:
What an excellent review, thank you. I'll get on that!

jasmussen committed:
Scope the is-selected color definition.

jasmussen commented:
There's definitely room to refactor and remove a ton of stuff. In this case I think the recently pushed change fixes it right up. If I'm right, the color-inheritance rule was to blame as it was attached to `.is-selected` rather than `.wp-block-navigation.is-selected`. So if you selected a parent, as in your case the Header, it would affect the placeholder. Shown here, a group with a black background, and the navigation inside is set to inherit the `currentColor` of borders and text:
![nav](https://user-images.githubusercontent.com/1204802/151343044-98c6563b-69c9-4b01-82e2-2a15bbfc70ef.gif)

☝️ that used to be broken, just like it was in your screen recording.

getdave reviewed (approved):
> There's definitely room to refactor and remove a ton of stuff.

No need in this PR. This work is great!

> ☝️ that used to be broken, just like it was in your screen recording.

Yep that's fixed it! 👏

jasmussen commented:
Thank you for the reviews and help!
